rust_icu_umsg/src/lib.rs

```diff
@@ -0,0 +1,472 @@
+///
+/// In general this is very brittle, and an API in a more modern lanugage, or a contemporary C++
+/// flavor would probably take a different route were the library to be written today.  The rust
+/// binding tries to make the API use a bit more palatable by requiring that the programmer
+/// explicitly specifies a type for each of the parameters to be passed into the formatter.
+///
+/// The supported types are not those of a full rust system, but rather a very restricted subset
+/// of types that MessageFormat supports:
+///
+/// | Type | Description |
+/// | ---- | ----------- |
+/// | Double | This is effectively rust type `f64`.  Any numeric parameter not specifically designated as different type, is always a double. See section below on Doubles. |
+/// | String | This is effectively [rust_icu_ustring::UChar].|
+/// | Integer | This is effectively `i32`, and is used |
+/// | Date | This is effectively [rust_icu_sys::UDate], and is used to format dates.  Depending on the date format requested in the pattern used in [UMessageFormat], the end result of date formatting could be one of a wide variety of [date formats](http://userguide.icu-project.org/formatparse/datetime).|
+///
+/// ## Double as numeric parameter
+///
+/// According to the [ICU documentation for
+/// `umsg_format`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/umsg_8h.html#a90a4b5fe778754e5da52f7c2e5fd6048):
+///
+/// > for all numeric arguments double is assumed unless the type is explicitly
+/// > integer (long).  All choice format arguments must be of type double.
+///
+/// ## Strings
+///
+/// We determined by code inspection that the string format must be `rust_icu_ustring::UChar`.
+///
+/// # Example use
+///
+/// ```
+/// use rust_icu_sys as sys;
+/// use rust_icu_common as common;
+/// use rust_icu_ustring as ustring;
+/// use rust_icu_uloc as uloc;
+/// use rust_icu_umsg::{self as umsg, message_format};
+/// # use rust_icu_ucal as ucal;
+/// # use std::convert::TryFrom;
+/// #
+/// # struct TzSave(String);
+/// # impl Drop for TzSave {
+/// #    // Restore the system time zone upon exit.
+/// #    fn drop(&mut self) {
+/// #        ucal::set_default_time_zone(&self.0);
+/// #    }
+/// # }
+///
+/// fn testfn() -> Result<(), common::Error> {
+/// # let _ = TzSave(ucal::get_default_time_zone()?);
+/// # ucal::set_default_time_zone("Europe/Amsterdam")?;
+///   let loc = uloc::ULoc::try_from("en-US")?;
+///   let msg = ustring::UChar::try_from(
+///     r"Formatted double: {0,number,##.#},
+///       Formatted integer: {1,number,integer},
+///       Formatted string: {2},
+///       Date: {3,date,full}",
+///   )?;
+///
+///   let fmt = umsg::UMessageFormat::try_from(&msg, &loc)?;
+///   let hello = ustring::UChar::try_from("Hello! Добар дан!")?;
+///   let _ = message_format!(&fmt);
+///   let result = umsg::message_format!(
+///     fmt,
+///     { 43.4 => Double },
+///     { 31337 => Integer},
+///     { hello => String},
```

Comment:
As a future improvement, we might want to support `&str` literals here.

Reply:
I am still not sure about this API... I guess we will see.